plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,300 @@
+     */
+    private final Map<MetricType, List<SearchQueryRecord>> topQueriesHistoryStores;
+    /**
+     * window start timestamp for each top queries collector
+     */
+    private final Map<MetricType, Long> topQueriesWindowStart;
+
+    /**
+     * Create the TopQueriesByLatencyService Object
+     *
+     * @param threadPool     The OpenSearch thread pool to run async tasks
+     */
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool) {
+        topQueriesStores = new HashMap<>();
+        topQueriesHistoryStores = new HashMap<>();
+        topQueriesWindowStart = new HashMap<>();
+        for (MetricType metricType : MetricType.allMetricTypes()) {
+            topQueriesStores.put(metricType, new PriorityBlockingQueue<>(topNSize, (a, b) -> SearchQueryRecord.compare(a, b, metricType)));
+            topQueriesHistoryStores.put(metricType, new ArrayList<>());
+            topQueriesWindowStart.put(metricType, -1L);
+        }
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Ingest the query data into in-memory stores
+     *
+     * @param record the record to ingest
+     */
+    public void addRecord(SearchQueryRecord record) {
+        for (MetricType metricType : record.getMeasurements().keySet()) {
+            this.threadPool.schedule(() -> {
+                if (!topQueriesStores.containsKey(metricType)) {
+                    return;
+                }
+                // add the record to corresponding priority queues to calculate top n queries insights
+                PriorityBlockingQueue<SearchQueryRecord> store = topQueriesStores.get(metricType);
+                checkAndResetWindow(metricType, record.getTimestamp());
+                if (record.getTimestamp() > topQueriesWindowStart.get(metricType)) {
+                    store.add(record);
+                    // remove top elements for fix sizing priority queue
+                    if (store.size() > this.getTopNSize()) {
+                        store.poll();
+                    }
+                }
+            }, delay, QueryInsightsSettings.QUERY_INSIGHTS_EXECUTOR);
+        }
+    }
+
+    private synchronized void checkAndResetWindow(MetricType metricType, Long timestamp) {
+        Long windowStart = calculateWindowStart(timestamp);
+        // reset window if the current window is outdated
+        if (topQueriesWindowStart.get(metricType) < windowStart) {
+            // rotate the current window to history store only if the data belongs to the last window
+            if (topQueriesWindowStart.get(metricType) == windowStart - windowSize.getMillis()) {
+                topQueriesHistoryStores.put(metricType, new ArrayList<>(topQueriesStores.get(metricType)));
+            } else {
+                topQueriesHistoryStores.get(metricType).clear();
+            }
+            topQueriesStores.get(metricType).clear();
+            topQueriesWindowStart.put(metricType, windowStart);
+        }
+    }
```

Comment:
Good point. Resetting window should only happen once per window - for most of the time, the window doesn't need to be reset. I think I can make the "check window size logic" unsynchronized, and do the synchronized reset only once per window.

The `LinkedBlockingQueue` + `drainTo` + `maintaining Top N` is similar to what I'm doing with the `PriorityBlockingQueue` + `maintaining top N`. let me further evaluate this.

Reply:
that's a good point and these record contains the query source which makes it bulky thus not limiting wait queue size can potentially cause memory issues. 
Also, it would be nice if you can limit the request making it this queue by introducing a setting like `minLatencyThreshold` and reject them straightaway, to avoid very low latency request waiting on this queue. Also, it would avoid unnecessary churning of your priority queue with low latency request getting added and popped for no reason.

Reply:
Thanks for the input! I have refactored the service code to make improve the multithreading efficiency. More specifcally, I added a `LinkedBlockingQueue` to cache the records and drain to different priority queues in a fixed interval. Also set a upper bound for the LinkedBlockingQueue to avoid memory issues.

I also added `AtomicReference` objects as cache of the current and last window top n records, and update those objects only at every interval. The read now all happens on those `AtomicReference` objects, which should make things more efficient.

Also refacored the code as froh suggested, to make a separate class for `TopQueries` with different `MetricType`, to avoid using too many HashMap<MetricTypes, ...> in the code.

Let me know what you think!